### PR TITLE
cylc checkpoint: new command

### LINKED ISSUE #1735: Explicit restart checkpointing?
~~At the moment cylc dumps suite state at every change, and keeps a rolling archive of the most recent states.  However, it's not very easy for users to determine which state file to restart from, if not the default (most recent), and you can't go far back either because the archive length is short (by default anyway).~~

[UPDATE: as of 6.11.0, the run DB has replaced the old state dump files, including for restarts.  As of 6.11.2, the DB automatically stores checkpoints representing the most recent suite state (continually updated) plus  before and after a reload, and at restart.  We can list available checkpoints with `cylc ls-checkpoints` and restart from any stored checkpoint.]

It might be useful to allow special checkpoint "tasks" in the graph to regularly ~~dump state~~ [write state checkpoints] at special points in the workflow that the user knows they might want to return to:

```
graph = foo => bar => :checkpoint: => baz => qux
```

This could be used to backtrack to an earlier point at the behest of some kind of diagnosis task, as an extreme form of automatic error recovery.

### PR DESCRIPTION
**Pull request by matthewrmshin**
New command to tell a suite to checkpoint itself.

Close #1735.

### CODE REVIEW

File: doc/cug.tex [hunk 1]
```diff
@@ -6118,6 +6118,21 @@ \subsubsection{Restart From Checkpoint}
 $ cylc restart --checkpoint=CHECKPOINT-ID SUITE
 \end{lstlisting}
 
+\subsubsection{Manual Checkpoints}
+
+You can use the \lstinline=cylc take-checkpoints= command to tell a running to
```
Comment (oliver-sanders) [inline doc/cug.tex]: `take-checkpoint` not `take-checkpoints`.

Comment (matthewrmshin) [inline doc/cug.tex]: Doc fixed.

File: doc/cug.tex [hunk 2]
```diff
@@ -6118,6 +6118,21 @@ \subsubsection{Restart From Checkpoint}
 $ cylc restart --checkpoint=CHECKPOINT-ID SUITE
 \end{lstlisting}
 
+\subsubsection{Manual Checkpoints}
+
+You can use the \lstinline=cylc take-checkpoint= command to tell a running to
```
Comment (hjoliver) [inline doc/cug.tex]: "...tell a running SUITE to ..."

File: bin/cylc-take-checkpoint [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        sys.exit(0)
+
+import cylc.flags
+from cylc.option_parsers import CylcOptionParser as COP
+from cylc.network.suite_command_client import SuiteCommandClient
+
+
+def main():
+    parser = COP(
+        __doc__, comms=True, multitask=True,
```
Comment (hjoliver) [inline bin/cylc-take-checkpoint]: `multitask=True` is wrong - it evokes the full task-matching documentation in the command help.

File: doc/cug.tex [hunk 4]
```diff
@@ -6118,6 +6118,21 @@ \subsubsection{Restart From Checkpoint}
...
+record a checkpoint at any time:
+
+\lstset{language=transcript}
+\begin{lstlisting}
+$ cylc take-checkpoint SUITE CHECKPOINT-NAME
+\end{lstlisting}
+
+The 2nd argument is a name you give to the checkpoint so you can easily
+identify when you need to use it.
```
Comment (hjoliver) [inline doc/cug.tex]: "identify IT when you need to use it"?

**Comment by hjoliver** [comment]
Maybe the command name should be just `checkpoint`, not `take-checkpoint` ("to checkpoint" something essentially means the same as "to take a checkpoint", I think... :grimacing: )

**Comment by matthewrmshin** [comment]
Comments addressed. Branch re-based and squashed.

**Pull request merged**